### PR TITLE
feat: support for passing folder when resolving schema

### PR DESCRIPTION
**Pull request by richtia**

### CODE REVIEW

File: src/gateway/demo/client_demo.py [hunk 1]
```diff
@@ -10,10 +10,9 @@
 
 # pylint: disable=fixme
 def get_customer_database(spark_session: SparkSession) -> DataFrame:
-    # TODO -- Support reading schema from multiple files.
-    location_customer = str(Path('../../../third_party/tpch/parquet/customer/part-0.parquet').absolute())
+    location_customer = str(Path('../../../third_party/tpch/parquet/customer').absolute())
```
Comment (EpsilonPrime) [inline src/gateway/demo/client_demo.py:13]: resolve will also make the path absolute but will eliminate the upward references as well

File: src/gateway/demo/client_demo.py [hunk 2]
```diff
@@ -10,10 +10,9 @@
...
 
-    return spark_session.read.parquet(location_customer,
+    return spark_session.read.parquet(location_customer + '/*.parquet',
```
Comment (EpsilonPrime) [inline src/gateway/demo/client_demo.py:15]: If your provide just the directory the parquet files will be discovered automatically.

File: src/gateway/converter/spark_to_substrait.py [hunk 3]
```diff
@@ -1,7 +1,9 @@
 # SPDX-License-Identifier: Apache-2.0
 """Routines to convert SparkConnect plans to Substrait plans."""
+import glob
```
Comment (EpsilonPrime) [inline src/gateway/converter/spark_to_substrait.py:3]: There's a glob method on a Path so you don't need this import.

File: src/gateway/converter/spark_to_substrait.py [hunk 4]
```diff
@@ -25,16 +27,21 @@
...
-
 DUCKDB_TABLE = "duckdb_table"
 
 
 def fetch_schema_with_adbc(path):
     """Fetch the arrow schema via ADBC."""
 
+    p = pathlib.Path(path)
+    folder = p.parent
+    file_paths = [str(folder / pathlib.Path(x)) for x in glob.glob(path)]
```
Comment (EpsilonPrime) [inline src/gateway/converter/spark_to_substrait.py:38]: You could wrap the right side with sorted() here (dropping the square brackets).

File: src/gateway/converter/spark_to_substrait.py [hunk 5]
```diff
@@ -25,16 +27,21 @@
...
+    file_paths.sort() # We sort the files because the later partitions don't have enough data to construct a schema.
+    file_path = file_paths[0]
+
     with adbc_driver_duckdb.dbapi.connect() as conn, conn.cursor() as cur:
         # TODO: Support multiple paths.
-        reader = pyarrow.parquet.ParquetFile(path)
+        reader = pyarrow.parquet.ParquetFile(file_path)
         cur.adbc_ingest(DUCKDB_TABLE, reader.iter_batches(), mode="create")
```
Comment (EpsilonPrime) [inline src/gateway/converter/spark_to_substrait.py:45]: This looks like it reads the file locally and ship it to the remote server.  When we separate the servers we will want reads to be done from the server's view (and not the client's).

File: src/gateway/demo/client_demo.py (hunk 2 above)
Comment (richtia) [reply src/gateway/demo/client_demo.py:15]: ahh okay...i can use the information from the read relation to determine that we'll want the parquet files.

File: src/gateway/converter/spark_to_substrait.py (hunk 3 above)
Comment (richtia) [reply src/gateway/converter/spark_to_substrait.py:3]: this is being used elsewhere now.

File: src/gateway/converter/spark_to_substrait.py [hunk 6]
```diff
@@ -402,7 +410,11 @@ def convert_read_data_source_relation(self, rel: spark_relations_pb2.Read) -> al
                 read=algebra_pb2.ReadRel(base_schema=schema,
                                          named_table=algebra_pb2.ReadRel.NamedTable(
                                              names=['demotable'])))
-        for path in rel.paths:
+        if pathlib.Path(rel.paths[0]).is_dir():
+            file_paths = glob.glob(f'{rel.paths[0]}/*{rel.format}')
```
Comment (EpsilonPrime) [inline src/gateway/converter/spark_to_substrait.py:414]: Optional:  You could use the / operator of paths to extend the path.

File: src/gateway/converter/spark_to_substrait.py [hunk 7]
```diff
@@ -25,16 +27,23 @@
     if_then_else_operation, greater_function, minus_function
 from gateway.converter.symbol_table import SymbolTable
 
-
 DUCKDB_TABLE = "duckdb_table"
 
 
-def fetch_schema_with_adbc(path):
+def fetch_schema_with_adbc(read_relation: spark_relations_pb2.Read) -> pyarrow.Schema:
```
Comment (EpsilonPrime) [inline src/gateway/converter/spark_to_substrait.py:33]: Optional:  Consider avoiding using the wire format as input to internal functionality (would you want to write a test for this function that required creating a read relation?).

**Review by EpsilonPrime — APPROVED** [review]

**Pull request merged**